pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -0,0 +1,261 @@
+				deleteFail(r, &ctx, obj)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "subnetset", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			log.Info("finalizers cannot be recognized", "subnetset", req.NamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func updateFail(r *SubnetSetReconciler, c *context.Context, o *v1alpha1.SubnetSet) {
+	r.setSubnetSetReadyStatusFalse(c, o)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, MetricResTypeSubnetSet)
+}
+
+func deleteFail(r *SubnetSetReconciler, c *context.Context, o *v1alpha1.SubnetSet) {
+	r.setSubnetSetReadyStatusFalse(c, o)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResTypeSubnetSet)
+}
+
+func updateSuccess(r *SubnetSetReconciler, c *context.Context, o *v1alpha1.SubnetSet) {
+	r.setSubnetSetReadyStatusTrue(c, o)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateSuccessTotal, MetricResTypeSubnetSet)
+}
+
+func deleteSuccess(r *SubnetSetReconciler, _ *context.Context, _ *v1alpha1.SubnetSet) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResTypeSubnetSet)
+}
+
+func (r *SubnetSetReconciler) setSubnetSetReadyStatusTrue(ctx *context.Context, subnetset *v1alpha1.SubnetSet) {
+	newConditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionTrue,
+			Message: "NSX SubnetSet has been successfully created/updated",
+			Reason:  "SubnetsReady",
+		},
+	}
+	r.updateSubnetSetStatusConditions(ctx, subnetset, newConditions)
+}
+
+func (r *SubnetSetReconciler) setSubnetSetReadyStatusFalse(ctx *context.Context, subnetset *v1alpha1.SubnetSet) {
+	newConditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionFalse,
+			Message: "NSX SubnetSet could not be created/updated",
+			Reason:  "SubnetNotReady",
+		},
+	}
+	r.updateSubnetSetStatusConditions(ctx, subnetset, newConditions)
+}
+
+func (r *SubnetSetReconciler) updateSubnetSetStatusConditions(ctx *context.Context, subnetset *v1alpha1.SubnetSet, newConditions []v1alpha1.Condition) {
+	conditionsUpdated := false
+	for i := range newConditions {
+		if r.mergeSubnetSetStatusCondition(ctx, subnetset, &newConditions[i]) {
+			conditionsUpdated = true
+		}
+	}
+	if conditionsUpdated {
+		r.Client.Status().Update(*ctx, subnetset)
+		log.V(1).Info("updated Subnet", "Name", subnetset.Name, "Namespace", subnetset.Namespace,
+			"New Conditions", newConditions)
+	}
+}
+
+func (r *SubnetSetReconciler) mergeSubnetSetStatusCondition(ctx *context.Context, subnetset *v1alpha1.SubnetSet, newCondition *v1alpha1.Condition) bool {
+	matchedCondition := getExistingConditionOfType(newCondition.Type, subnetset.Status.Conditions)
+
+	if reflect.DeepEqual(matchedCondition, newCondition) {
+		log.V(2).Info("conditions already match", "New Condition", newCondition, "Existing Condition", matchedCondition)
```

Comment:
Don't use space to connect multi words in the key of structured logging, otherwise it could be hard to differentiate key and value in some cases and hard to use tool to extract key and values. Some suggestions about logging:
https://github.com/tnqn/code-review-comments#logging